modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteTableScan.java

```diff
@@ -44,7 +44,12 @@
      */
     public IgniteTableScan(RelInput input) {
         super(changeTraits(input, IgniteConvention.INSTANCE));
-        sourceId = ((Number)input.get("sourceId")).longValue();
+
+        Object srcIdObj = input.get("sourceId");
+        if (srcIdObj != null)
+            sourceId = ((Number)srcIdObj).longValue();
+        else
+            sourceId = -1;
```

Comment:
i put _assert false_ here and no tests are failed, have no ideas what is it code for, can u explain or return old functionality ?

Reply:
A sourceId is serialised if it is not equal to -1 (see `IgniteTableScan#explainTerms0()`). So when deserialising we could get NPE here, if the node was serialised with `sourceId=-1`. It's just a matter of time